pkg/daemon/certificate_writer.go

```diff
@@ -102,37 +118,74 @@ func (dn *Daemon) syncControllerConfigHandler(key string) error {
 	// Also annotate the latest config we've seen, so as to not write unnecessarily
 	currentNodeControllerConfigResource := dn.node.Annotations[constants.ControllerConfigResourceVersionKey]
 
-	if currentNodeControllerConfigResource != controllerConfig.ObjectMeta.ResourceVersion {
+	if currentNodeControllerConfigResource != controllerConfig.ObjectMeta.ResourceVersion || controllerConfig.Annotations[ctrlcommon.ServiceCARotateAnnotation] == ctrlcommon.ServiceCARotateTrue {
 		pathToData := make(map[string][]byte)
-
 		kubeAPIServerServingCABytes := controllerConfig.Spec.KubeAPIServerServingCAData
 		cloudCA := controllerConfig.Spec.CloudProviderCAData
 		pathToData[caBundleFilePath] = kubeAPIServerServingCABytes
 		pathToData[cloudCABundleFilePath] = cloudCA
+		var data []byte
+		var err error
+		var cmErr error
+		var cm *corev1.ConfigMap
 
-		for bundle, data := range pathToData {
-			if !strings.HasSuffix(string(data), "\n") {
-				bString := string(data) + "\n"
-				data = []byte(bString)
-			}
-			if Finfo, err := os.Stat(bundle); err == nil {
-				var mode os.FileMode
-				Dinfo, err := os.Stat(filepath.Dir(bundle))
-				if err != nil {
-					mode = defaultDirectoryPermissions
-				} else {
-					mode = Dinfo.Mode()
-				}
-				// we need to make sure we honor the mode of that file
-				if err := writeFileAtomically(bundle, data, mode, Finfo.Mode(), -1, -1); err != nil {
-					return err
-				}
+		if controllerConfig.Annotations[ctrlcommon.ServiceCARotateAnnotation] == ctrlcommon.ServiceCARotateTrue && dn.node.Annotations[constants.ControllerConfigSyncServerCA] != controllerConfig.Annotations[ctrlcommon.ServiceCARotateAnnotation] {
```

Comment:
suggestion: Move this equality test into a separate function called, e.g., `shouldRotateCerts()` or something along those lines to be more descriptive.

We check for these conditions twice, so I think it makes sense to give this check a more descriptive name.